lib/ret/storage.ex

```diff
@@ -61,19 +72,62 @@ defmodule Ret.Storage do
   end
 
   def fetch(id, key) when is_binary(id) and is_binary(key) do
-    fetch_blob(id, key, @expiring_file_path)
+    fetch_blob(id, key, expiring_file_path())
   end
 
   def fetch(%OwnedFile{owned_file_uuid: id, key: key}) do
-    fetch_blob(id, key, @owned_file_path)
+    fetch_blob(id, key, owned_file_path())
+  end
+
+  def fetch(%CachedFile{} = cached_file) do
+    fetch_at(cached_file, Timex.now() |> Timex.to_naive_datetime() |> NaiveDateTime.truncate(:second))
+  end
+
+  defp maybe_bump_accessed_at(%CachedFile{accessed_at: accessed_at} = cached_file, time) do
+    one_day_ago = Timex.shift(time, days: -1)
+
+    if accessed_at < one_day_ago do
+      cached_file |> Ecto.Changeset.change(accessed_at: time) |> Ret.Repo.update()
+    end
+  end
+
+  def fetch_at(%CachedFile{file_uuid: id, file_key: key} = cached_file, time) do
+    maybe_bump_accessed_at(cached_file, time)
+
+    case fetch_blob(id, key, cached_file_path()) do
+      {:ok, meta, stream} ->
+        {:ok, meta, stream}
+
+      {:error, _} ->
+        migrate(%{cached_file: cached_file})
```

Comment:
As we discussed, we should probably put a lock around this migration, or at least test to ensure that multiple simultaneous requests to the same file don't cause issues.